### ISSUE TITLE
Named parameter completions insert duplicated text when a value has already been provided

### ISSUE THREAD

**Issue by eseidel** [Dart-Code/Dart-Code#242]
I run into this occasionally.  I believe it's specific to named parameters:
![screen shot 2017-01-05 at 9 57 04 pm](https://cloud.githubusercontent.com/assets/478541/21708978/15bfdb2e-d392-11e6-94ea-13e3bcffca14.png)
Hit tab:
![screen shot 2017-01-05 at 9 57 10 pm](https://cloud.githubusercontent.com/assets/478541/21708986/1781e3ee-d392-11e6-9ce9-b9bc162ed0fc.png)

**Comment by DanTup** [comment]
By "occasionally", do you mean it's intermittent, or that it's reliably reproducible but only for certain code?

Do you happen to have the analyzer logging turned on? You can set `dart.analyzerLogFile` for a reduced version from Dart Code or `dart.analyzerInstrumentationLogFile` but the built-in one (I tend to run with the Dart Code one on all the time so if anything weird happens I don't need to try and repro, though it can get big and need deleting now and then!).

I made some changes to this recently (due to a break in the Code API), it's very possible I just tested this badly; I'll take a look when I can. If you can repro in a single Dart file, that would be handy!

Do you have `dart.insertArgumentPlaceholders` set in your config? Looking at code, I wonder if this happens when you already have the value (eg. `: 10.0`) for the argument you're completing?

**Comment by eseidel** [comment]
These are my user settings:
```
// Place your settings in this file to overwrite the default settings
{
        "editor.formatOnType": true,
            "editor.formatOnSave": true,

}
```

**Comment by eseidel** [comment]
It seems to be reproducible for a certain set of code.  But I don't know what a reduction would look like.  I tried to produce one (from scratch) just now and failed.

**Comment by DanTup** [comment]
Thanks for trying, I'll do some digging. I don't suppose it's a public repro where you can reproduce it?

Enabling one of the logs might be useful to see what the edit that comes back from the analysis server looks like.

**Comment by eseidel** [comment]
Here is the exact line which I filed about:
https://github.com/eseidel/lolsim/blob/e604b1fe4e433d6b4959d576628fca0b1cbbca0e/lib/lolsim.dart#L634
And yes, it reproduces reliably trying to tab-complete the 'magicDamage' argument name.

**Comment by eseidel** [comment]
Is "dart.analyzerLogFile" a user setting in Visual Studio Code?  I'm not sure where to put that option.  It doesn't auto-complete for me when I try to add it to User Settings.

**Comment by eseidel** [comment]
Yeah, when I tried adding dart.analyzerLogFile to my user settings in visual code, it tells me property unexpected.

**Comment by DanTup** [comment]
Cool; I'll try and repro it from there.

> Is "dart.analyzerLogFile" a user setting in Visual Studio Code?
> Yeah, when I tried adding dart.analyzerLogFile to my user settings in visual code, it tells me property unexpected.

It's a "hidden" property because it's for debugging so most users won't be interested in it. You might get a warning in the editor, but it should still work. You'll need to reload the project (there's a "Reload Window" command in Code that'll do this) after changing and then the file should be created.

I don't know how useful it'll be; we truncate the JSON and I think the completion responses are huge (they return lots and the IDE filters).

Don't worry about the logging too much - if the issue can be reproduced from the repo above I should be able to step through with a debugger and figure it out - the logs might've been useful if it wasn't reliably reproducible.

**Comment by DanTup** [comment]
This did indeed turn out to be caused by the value already existing. It can repro'd with some simple code:

```dart
method2({bool danny}) {}
method4() {
  method2(da: "test");
}
```

If you complete after `da` you get the duplication. This is because we normally insert placeholders when completing (so you'd get `danny: danny` with the second danny as a placeholder you can type over) but I'd overlooked that the value might already exist (in which case the completion won't end with `: `).

To fix this, we won't insert the value placeholders unless the completion ends with `: `. This seems to behave correctly in all scenarios I came up with while testing, but do shout if you see it after upgrading.

**Comment by DanTup** [comment]
@eseidel This fix is now live (v0.14.9). Please give it a test and let me know if you have any more issues. Thanks!
